File: LeetCode_Monthly_Challenge/June/Python3/Week_2/insert_delete_getrandom_O(1).py
"""
PROBLEM:
Design a data structure that supports all following operations in average O(1) time.
insert(val): Inserts an item val to the set if not already present.
remove(val): Removes an item val from the set if present.
getRandom: Returns a random element from current set of elements. Each element must have the same probability of being returned.
"""


class RandomizedSet:
    def __init__(self):
        self.map, self.arr={}, []

    def insert(self, val: int) -> bool:
        if val not in self.map:
            self.map[val]=len(self.arr)
            self.arr.append(val)
            return True
        return False

    def remove(self, val: int) -> bool:
        if val in self.map:
            i, last = self.map[val], self.arr[-1]
            self.arr[i], self.map[last] = last, i
            self.arr.pop()
            self.map.pop(val)
            return True
        return False
            
    def getRandom(self) -> int:
        return random.choice(self.arr);
        
# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: LeetCode_Monthly_Challenge/June/C++/Week_2/Insert Delete Getrandom O(1).cpp
/*
PROBLEM:
Design a data structure that supports all following operations in average O(1) time.
insert(val): Inserts an item val to the set if not already present.
remove(val): Removes an item val from the set if present.
getRandom: Returns a random element from current set of elements. Each element must have the same probability of being returned.
*/


class RandomizedSet {
public:
    unordered_map<int,int> map;
    vector<int> v;
    RandomizedSet() {   
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(map.count(val))
            return false;
        map[val]=v.size();
        v.push_back(val);
        return true;
        
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(!map.count(val)) 
            return false;
        int i=map[val];
        map[v.back()]=i;
        map.erase(val);
        swap(v[i],v.back());
        v.pop_back();
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return v[rand()%v.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */
